PR: Improve UI/UX of the Run plugin configuration widgets

ccordoba12 opened the pull request:
## Description of Changes

- The new architecture for the Run plugin (added in #17467) is quite powerful and flexible. However, it's not easy to understand how to make use of it.
- This PR aims to improve that by
    - Adding default execution parameters to our config system. That will show them in the Preferences page for Run, which will allow users to easily modify them. In addition, that page won't appear empty, which is a bit odd.
    - Simplifying significantly the UI of the dialog that appears when you go to `Run > Open run settings`. In addition, that dialog was modified so that it allows to only set run settings per file (just as in Spyder 5). That makes it simpler to understand and leaves creating global execution configs to the Run entry in Preferences.
- Add a new `CollapsibleWidget` widget, which is based on `QCollpasible` from SuperQt, and use it in `RunDialog`. That required increasing our minimal constraint on SuperQt to `>=0.6.2`.

### Visual changes

#### Configuration per file

* Before

    ![imagen](https://github.com/spyder-ide/spyder/assets/365293/26515b4e-1b7d-4b11-8ddd-20b9ea515d90)

* After

    ![imagen](https://github.com/spyder-ide/spyder/assets/365293/4e294cbe-d697-49bd-919b-dca43a04f307)

#### Preferences page

* Before

    ![imagen](https://github.com/spyder-ide/spyder/assets/365293/f898e6e0-3bd0-4005-975f-40c7b024eb92)

* After

    ![imagen](https://github.com/spyder-ide/spyder/assets/365293/822dd54e-3ad0-49a1-876a-a9faccad57ff)

#### Dialog to create global run configurations

* Before

    ![imagen](https://github.com/spyder-ide/spyder/assets/365293/83cf96d5-f303-425e-a39f-812274652045)

* After

   ![imagen](https://github.com/spyder-ide/spyder/assets/365293/ec83d2c5-026b-4d20-a2bf-af5001659082)

### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #21878
Fixes #20700
Fixes #20569

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @ccordoba12 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @ccordoba12! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







































* In the file [`spyder/plugins/run/models.py`](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/models.py):

> [Line 454:14](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/models.py#L454): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

* In the file [`spyder/plugins/run/tests/test_run.py`](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/tests/test_run.py):

> [Line 644:46](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/tests/test_run.py#L644): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 816:9](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/tests/test_run.py#L816): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation

* In the file [`spyder/plugins/run/widgets.py`](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/widgets.py):

> [Line 819:14](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/plugins/run/widgets.py#L819): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'



* In the file [`spyder/utils/icon_manager.py`](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/utils/icon_manager.py):

> [Line 381:80](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/utils/icon_manager.py#L381): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 79 characters)
> [Line 382:80](https://github.com/spyder-ide/spyder/blob/0a2a4b38df77c9fb799f90d8c5f15eda32542a03/spyder/utils/icon_manager.py#L382): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 79 characters)

ccordoba12 commented:
@dalthviz, this is ready for review. Please let me know issues/shortcomings you find with the new implementation.

dalthviz commented:
Gave this an initial local check on Windows and saw some behaviors/interactions that felt kind of strange (the last couple of things are more likely bugs):

* Information tooltip refering to the `Ok` button at first glance make me think that it was refering to the `Save globally` button instead of the `OK` button at the bottom of the dialog itself. Not sure if it was discussed before, but now that I'm trying the dialog is kind of strange to have buttons in different places of the dialog: ![image](https://github.com/spyder-ide/spyder/assets/16781833/5936730f-df6d-4af4-bd54-1a70db677ac2) Also, the `Delete` button refers to a global delete of the config right?

* Being able to trigger a file run config delete when I haven't save anything yet (even a confirmation dialog appears and then the run dialog gets closed): ![run_1](https://github.com/spyder-ide/spyder/assets/16781833/29a5fb6c-0f42-4002-b9bd-bbad58f02771)

* Triggering a `Save globally` without entering a name/leaving the default one triggers a confirmation dialog before showing and red circle x over the config name lineedit: ![run_2](https://github.com/spyder-ide/spyder/assets/16781833/f6355a85-0ab8-46f0-93e8-b2d1fb4641b4)

* Trying to save a new global config without a name changes the lineedit placeholder to a message indicating a name is needed. Maybe to make things more clear putting the red circle x over the lineedit: 
![image](https://github.com/spyder-ide/spyder/assets/16781833/842d68f7-1fa0-4064-9bcf-e263ef008b96)

* Seems like trying to modify a global config from the run config dialog ends up deleting the global config?:

![run_3](https://github.com/spyder-ide/spyder/assets/16781833/567f8b3a-c3b4-4fe4-8e5c-c647b30f3415)

* A traceback when using `Reset changes` and `Edit selected` buttons on the Run preference page:

```python-traceback
Traceback (most recent call last):
  File "E:\Acer\Documentos\Spyder\Spyder otros\carlos\spyder\spyder\plugins\run\confpage.py", line 251, in <lambda>
    lambda checked: self.params_table.show_editor()
  File "E:\Acer\Documentos\Spyder\Spyder otros\carlos\spyder\spyder\plugins\run\confpage.py", line 111, in show_editor
    (extension, context, params) = model[index]
  File "E:\Acer\Documentos\Spyder\Spyder otros\carlos\spyder\spyder\plugins\run\models.py", line 540, in __getitem__
    tuple_index = self.params_index[index]
KeyError: -1
```

Other than that seems like things are working as expected but probably having more people testing this could be worthy (seems like there could be some interactions that I haven't checked yet :thinking:)

jitseniesen commented:
Just looking at the pictures, I would remove the heading "Runner settings" in the dialog windows. As a user, it is not clear why the working directory settings are separate from the runner settings. This will also simplify the dialog by removing one level of nesting.

ccordoba12 commented:
To @dalthviz:

> Information tooltip refering to the Ok button at first glance make me think that it was refering to the Save globally button instead of the OK button at the bottom of the dialog itself. Not sure if it was discussed before, but now that I'm trying the dialog is kind of strange to have buttons in different places of the dialog

We discussed this with @conradolandia and agree with you. That's why I decided to remove "Save globally" and move the Delete button to the dialog button box at the bottom.

To help understand users where to set global configurations, I added a new entry called like that to the Run menu:

![imagen](https://github.com/spyder-ide/spyder/assets/365293/af8d8713-501a-4ed8-966c-3bd40fad2c0c)

> Also, the Delete button refers to a global delete of the config right?

No, it refers to deleting file configurations only. To clarify that, I changed the behavior of that button so that it's disabled for default or global configurations.

> Being able to trigger a file run config delete when I haven't save anything yet (even a confirmation dialog appears and then the run dialog gets closed)

It should be fixed now by the change I mentioned above.

> Triggering a Save globally without entering a name/leaving the default one triggers a confirmation dialog before showing and red circle x over the config name lineedit

It's no longer a problem.

> Trying to save a new global config without a name changes the lineedit placeholder to a message indicating a name is needed. Maybe to make things more clear putting the red circle x over the lineedit

That's a great suggestion to make things consistent. I implemented it in one of my last commits.

> Seems like trying to modify a global config from the run config dialog ends up deleting the global config?

It shouldn't be a problem anymore.

> A traceback when using Reset changes and Edit selected buttons on the Run preference page

Thanks for noticing it. It should be fixed now.

----

To @jitseniesen:

> Just looking at the pictures, I would remove the heading "Runner settings" in the dialog windows

Great suggestion! I did that in one of my last commits (see updated screenshots in the OP).

ccordoba12 commented:
In addition to fixing the problems found in review, I took the opportunity to improve the design of the Run page in Preferences (see the updated OP).

dalthviz commented:
Gave another check and I think the difference between the local vs global configs is clearer now :+1: However, there are a couple of things that seem like bugs/I'm still not sure about:

* Cloning/duplicationg a global config and then removing the selected config removes all the created global configs: 
![run_21](https://github.com/spyder-ide/spyder/assets/16781833/b9adc7c4-e470-4fe0-8c63-926ae22ed5a9)

* Create a new local config, open the dialog again, try to modify the config without changing the name lineedit. The circular red x appears. Following that, try to set as preset configuration the default from the combobox (or use the reset button), still unable to close the dialog via the `Ok` button

![run_22](https://github.com/spyder-ide/spyder/assets/16781833/1e2da7d7-7178-43e3-829b-ac600fccbe27)

* Open the run dialog and choose a different preset or even just open and try to close it via the `Ok` button the dialog will be kept open due to the name:

![run_23](https://github.com/spyder-ide/spyder/assets/16781833/07bfa2a5-3adf-404c-8a50-759ec5be5c85)

So my guess is that it is not possible to override an existing local config? In fact, as things are right now, is not possible to select one of the already created configs/presets without creating a new config, right? If that is the case, probably more feedback about the error needs to be provided. Some ideas:
* Add some sort of message somewhere indicating that a new name is needed
* When showing the collapsible section, remove the text along side showing the circle red x and move focus to the line edit to point people not only that something is invalid but that they need to take action to overcome it

From the checks I have been doing, I would say that from the run dialog it would be nice to support the following operations:
* Select a config from the presets and use it without having to do any modification (without having to deal with the `Custom configuration` section)
* Be able to select a preset and change a specific config without having to create a new one
* Be able to select a preset and use it as a base for a new config with a different name

Not sure how much of the these things are feasible/fall under the scope you had for this PR/the general functionality of the run dialog @ccordoba12 but sharing the ideas that came to mind while checking this

ccordoba12 commented:
Thanks @dalthviz for the additional review! About your comments:

> Cloning/duplicationg a global config and then removing the selected config removes all the created global configs

Good catch! It should be fixed now.

> Create a new local config, open the dialog again, try to modify the config without changing the name lineedit. The circular red x appears.
> Open the run dialog and choose a different preset or even just open and try to close it via the Ok button the dialog will be kept open due to the name
> So my guess is that it is not possible to override an existing local config?

These problems were an oversight on my side. The thing is I added a validation so that users can't introduce repeated config names in the run dialog. But I forgot to skip it when the config name and the one shown in the preset combobox are the same. So, please test again.

> Select a config from the presets and use it without having to do any modification (without having to deal with the Custom configuration section)

This should be fixed now.

> Be able to select a preset and change a specific config without having to create a new one

This is not possible anymore, i.e. with the changes done here. I think it was before, but then it gave users the odd impression that settings were not saved (see issue #20700, which was reported several times).

However, to not force users to provide a name the first time they change global parameters, I decided to set `Custom for this file` as the config name in that case. But for any other change to those parameters, they'll need to set a name of their own.

> Be able to select a preset and use it as a base for a new config with a different name

This should be possible now. Please check again.

dalthviz commented:
Gave this another check and seems like things are now working as expected. Will give the code another check but for the most part I would say that this looks good to me :+1:

dalthviz commented on spyder/app/tests/test_mainwindow.py:
```diff
@@ -946,8 +955,10 @@ def test_shell_execution(main_window, qtbot, tmpdir):
-@pytest.mark.skipif(sys.platform.startswith('linux'),
-                    reason="Fails frequently on Linux")
+@pytest.mark.skipif(
+    sys.platform.startswith('linux') and running_in_ci(),
+    reason="Fails frequently on Linux"
```
So this is working on linux locally but not on CI, right?

```suggestion
    sys.platform.startswith('linux') and running_in_ci(),
    reason="Fails frequently on Linux CI"
```

dalthviz commented on spyder/app/utils.py:
```diff
@@ -187,6 +187,15 @@ def qt_message_handler(msg_type, msg_log_context, msg_string):
         # This is shown when expanding/collpasing folders in the Files plugin
         # after spyder-ide/spyder#
         "QFont::setPixelSize: Pixel size <= 0 (0)",
+        # These warnings are shown uncollapsing CollapsibleWidget
+        "QPainter::begin: Paint device returned engine == 0, type: 2",
+        "QPainter::save: Painter not active",
+        "QPainter::setPen: Painter not active",
+        "QPainter::setWorldTransform: Painter not active",
+        "QPainter::setOpacity: Painter not active",
+        "QFont::setPixelSize: Pixel size <= 0 (-3)",
+        "QPainter::setFont: Painter not active",
+        "QPainter::restore: Unbalanced save/restore",
```
Should we create an issue to determine the exact reason for this warnings?

ccordoba12 replied:
These warnings come from the SuperQt implementation of `QCollapsible`. So, we can't address them here.

dalthviz commented on spyder/app/utils.py:
```diff
@@ -358,6 +367,16 @@ def create_window(WindowClass, app, splash, options, args):
+    #   height or position) that other widgets can need to position relative to
+    #   it.
+    # * **DO NOT** use it to access other plugins functionality through it.
+    app.main_window = main
```
Rather than giving access to the whole main window shouldn't be created specific accessors? So, leaving this as `app._main_window = main` and creating a set of functions to get the values needed so `get_mainwindow_width`, `get_mainwindow_height`, `get_mainwindow_position`, etc?

ccordoba12 replied:
That's a very good idea! Thanks @dalthviz for the suggestion 👍🏽 

I'll implement it before merging.

dalthviz commented on spyder/config/main.py:
```diff
@@ -643,7 +643,6 @@
         ('run', [
             'breakpoints',
             'configurations',
-            'defaultconfiguration',
```
There is no need to bump the config version, right?

ccordoba12 replied:
This config option wasn't used anywhere. That's why I didn't bump `CONF_VERSION`.

dalthviz commented on spyder/plugins/run/container.py:
```diff
@@ -88,6 +90,12 @@ def setup(self):
 
+        self.create_action(
+            RunActions.GlobalConfigurations,
+            _("Global configurations"),
```
Maybe it could be worthy to add a menubar shortcut to this new global action too? So:
```suggestion
            _("&Global configurations"),
```

dalthviz reviewed (approved):
Thanks @ccordoba12 ! Left a couple of questions/comments but this LGTM :+1:

Also, on a last manual check, I noticed that seems like you can create a global config with the same name of an already create per file config:

![run_config](https://github.com/spyder-ide/spyder/assets/16781833/c389908f-3426-4353-9f7b-a4b657313356)

Not sure if this could be thought as a bug or not but letting you know. I'm approving this so feel free to merge if you think is ready @ccordoba12

ccordoba12 commented:
> Also, on a last manual check, I noticed that seems like you can create a global config with the same name of an already create per file config

Yep, that's correct. Since global configs take precedence over file ones, I think that validation is not appropriate and so I didn't implement it. Furthermore, it could be too restrictive to users given that they can create as many file configs as they want for as many files as possible.

The good thing is that there a subtle clue in `RunDialog` about which configs are global and which ones belong to the file: the `Delete` button. If it's enabled, it means we're dealing with a file config; if not, it means the config is global.

If users let us know that's not enough, we can think about how to disambiguate them. But for now, I think that's ok.

ccordoba12 commented:
One last note: this PR also contains a couple of small fixes to our tests to deal with the Numpy 2.0 release (which is causing the failures in master).

Matthew-Copping commented:
After modifying the default behaviour and restarting spyder the original default configuration reappears along side my changed version. As a result my changes are not used. 

Is this expected behaviour? If so how do i change the default behaviour?

![image](https://github.com/user-attachments/assets/44496efa-1789-4270-9c3e-d2c6a265f505)

ccordoba12 commented:
> Is this expected behaviour? If so how do i change the default behaviour?

No, it's not. Please open a new issue about it and add to it a small video showing what you did to better understand your problem.

CAM-Gerlach commented:
#### Configuration per file

    | Before | After |
    | - | - |
    | ![imagen](https://github.com/spyder-ide/spyder/assets/365293/26515b4e-1b7d-4b11-8ddd-20b9ea515d90) | ![imagen](https://github.com/spyder-ide/spyder/assets/365293/4e294cbe-d697-49bd-919b-dca43a04f307)


#### Preferences page

    | Before | After |
    | - | - |
    | ![imagen](https://github.com/spyder-ide/spyder/assets/365293/f898e6e0-3bd0-4005-975f-40c7b024eb92) | ![imagen](https://github.com/spyder-ide/spyder/assets/365293/822dd54e-3ad0-49a1-876a-a9faccad57ff)


#### Global run configuration

    | Before | After |
    | - | - |
    | ![imagen](https://github.com/spyder-ide/spyder/assets/365293/83cf96d5-f303-425e-a39f-812274652045) | ![imagen](https://github.com/spyder-ide/spyder/assets/365293/ec83d2c5-026b-4d20-a2bf-af5001659082)
